### PR TITLE
OCPEDGE-1430: feat(deployment): install kubesan csi via make deploy

### PR DESCRIPTION
**Pull request by qJkee**
Add kubesan manifests to the repo to make it easier to install via regular make deploy command

### CODE REVIEW

File: config/kubesan/kubernetes/csi/node-plugin.yaml [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+              mountPropagation: Bidirectional # for Filesystem volume mounting
+              mountPath: /var/lib/kubelet
+            - name: socket-dir
+              mountPath: /run/csi
+            - name: dev
+              mountPath: /dev
+            - name: run-kubesan-nbd
+              mountPath: /run/kubesan/nbd
+          securityContext:
+            privileged: true
```
Comment (jaypoulz) [inline config/kubesan/kubernetes/csi/node-plugin.yaml:60]: I don't know enough about trampoline pods to advise if this is a risk, but I remember that daemonsets are usually the source of privilege trampolines in the cluster. Maybe you linked me to this, but it's probably worth double check that we're using the recommendations explained in https://www.youtube.com/watch?v=qRo1Qw_Hr2A

I think this daemonset also needs cpu architecture boundaries as well.

File: config/kubesan/kubernetes/manager/cluster.yaml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        app.kubernetes.io/name: kubesan-cluster-controller-manager
+        app.kubernetes.io/part-of: ksan-storage
+        app.kubernetes.io/component: cluster-controller-manager
+      annotations:
+        target.workload.openshift.io/management: '{"effect": "PreferredDuringScheduling"}'
+    spec:
+      serviceAccountName: cluster-controller-manager-kubesan
+      terminationGracePeriodSeconds: 10
+      hostPID: true
+      containers:
```
Comment (jaypoulz) [inline config/kubesan/kubernetes/manager/cluster.yaml:27]: same comment as above related to nodeSelector or affinity boundaries to prevent unsafe scheduling

File: config/kubesan/kubernetes/manager/node.yaml [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        app.kubernetes.io/name: kubesan-node-controller-manager
+        app.kubernetes.io/part-of: ksan-storage
+        app.kubernetes.io/component: node-controller-manager
+      annotations:
+        target.workload.openshift.io/management: '{"effect": "PreferredDuringScheduling"}'
+    spec:
+      serviceAccountName: node-controller-manager-kubesan
+      terminationGracePeriodSeconds: 10
+      hostPID: true
+      containers:
```
Comment (jaypoulz) [inline config/kubesan/kubernetes/manager/node.yaml:27]: arch affinity/selector

File: config/kubesan/openshift/patch_priority_class.yaml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
+- op: add
+  path: /spec/template/spec/priorityClassName
+  value: openshift-user-critical
```
Comment (jaypoulz) [inline config/kubesan/openshift/patch_priority_class.yaml:5]: Is this the standard priority class we use for all storage operators?

File: config/kubesan/openshift/scc.yaml [hunk 5]
```diff
@@ -0,0 +1,37 @@
+# SPDX-License-Identifier: Apache-2.0
+
+---
+apiVersion: security.openshift.io/v1
+kind: SecurityContextConstraints
+metadata:
+  name: kubesan
+allowPrivilegedContainer: true
```
Comment (jaypoulz) [inline config/kubesan/openshift/scc.yaml:8]: I'm assuming we need root because the code that manages system volumes needs it?

File: config/manager/manager.yaml [hunk 6]
```diff
@@ -1,75 +1,52 @@
...
       annotations:
-        kubectl.kubernetes.io/default-container: manager
+        target.workload.openshift.io/management: '{"effect": "PreferredDuringScheduling"}'
       labels:
-        control-plane: controller-manager
+        app.kubernetes.io/name: ksan-storage-operator
+        app.kubernetes.io/part-of: ksan-storage
+        app.kubernetes.io/component: ksan-storage-operator
     spec:
-      # TODO(user): Uncomment the following code to configure the nodeAffinity expression
```
Comment (jaypoulz) [inline config/manager/manager.yaml:31]: Hey! an example of how you should handle architectures :D
I suspect we want to set boundaries for amd64 at the minimum.

File: config/default/kustomization.yaml [hunk 7]
```diff
@@ -1,12 +1,12 @@
...
-namespace: ksan-system
+namespace: openshift-ksan-storage-operator
 
 # Value of this field is prepended to the
 # names of all resources, e.g. a deployment named
 # "wordpress" becomes "alices-wordpress".
 # Note that it should also match with the prefix (text before '-') of the namespace
 # field above.
-namePrefix: ksan-
+#namePrefix: ksan-
```
Comment (jaypoulz) [inline config/default/kustomization.yaml:9]: This confuses me because it's set here: https://github.com/openshift/ksan-operator/pull/10/files#diff-14effe9ea1a3ddb6e419e627a3501a6c42aa2dcbc76803923e9aff2438e6ee34R14

**Review by jaypoulz — COMMENTED** [review]
I've left you comments that fall into my domain of experience with operators. Overall, I don't feel like any changes are required, but I don't have the domain-specific knowledge to tell you if this operator is well-defined, follows best practices, and is built with security in mind. Overall, I'm fine with moving forward with this as a baseline, but I think we may want to have a team discussion around how we want to handle auditing this project for security vulnerabilities and usability issues.

File: config/default/kustomization.yaml (hunk 7 above)
Comment (jeff-roche) [reply config/default/kustomization.yaml:9]: I think in this case we just don't want it in the default, we primarily want it as part of the manager configuration.

File: config/kubesan/kubernetes/csi/controller-plugin.yaml [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+      app.kubernetes.io/name: kubesan-csi-controller-plugin
+  template:
+    metadata:
+      labels:
+        app.kubernetes.io/name: kubesan-csi-controller-plugin
+        app.kubernetes.io/part-of: ksan-storage
+        app.kubernetes.io/component: csi-controller-plugin
+      annotations:
+        target.workload.openshift.io/management: '{"effect": "PreferredDuringScheduling"}'
+    spec:
```
[the remark this answers is not in the document]
Comment (jeff-roche) [reply config/kubesan/kubernetes/csi/controller-plugin.yaml:29]: fwiw, we don't do this for lvms either

File: config/kubesan/kubernetes/csi/node-plugin.yaml (hunk 1 above)
Comment (jeff-roche) [reply config/kubesan/kubernetes/csi/node-plugin.yaml:60]: I think we have to set it to privileged in order to access the ksan binaries on the host

File: config/kubesan/openshift/patch_priority_class.yaml (hunk 4 above)
Comment (jeff-roche) [reply config/kubesan/openshift/patch_priority_class.yaml:5]: afaik, yes

File: config/manager/manager.yaml (hunk 6 above)
Comment (jeff-roche) [reply config/manager/manager.yaml:31]: afaik kubesan supports arm64 and amd64 currently. I don't know the long term plan for architectures though. If we plan to support all 4, we don't need this block. If we only plan to support arm and amd, then it makes sense to add it imho

Comment (jeff-roche) [reply config/manager/manager.yaml:31]: @brandisher do you know what the arch boundaries we plan to support are?

**Comment by jeff-roche** [comment]
I'm confused in general about the `config/kubesan` folder here. My understanding is that we wanted to do a bundle approach and I'd expect to see the manifests for kubesan in the openshift/kubesan repo so they are built into the kubesan container image. We would then create a bundle image that would pull in kubesan and start it up based on the manifests from kubesan. It seems like we are overloading this config repo with both operator and csi-driver which _feels_ like overkill to me and a too monolithic.

File: config/manager/manager.yaml (hunk 6 above)
Comment (DanielFroehlich) [reply config/manager/manager.yaml:31]: we need to support amd64 first and on inital release (Dev Preview).
arm64 is second prio and could be added post-ga. Also fine to add it right away if we can get it kind of for free with little effort or delays.
currently no long term plans for ppc64 or s390, but this might change based on customer demand.

**Comment by jaypoulz** [comment]
After speaking with Bulat - I'm of the opinion that we capture the two main reservations on this PR as tech debt that can be addressed at a later time so that we can maintain focus on meeting our milestones.

Specifically, these are the following points I've understood from the discussion.

- We would like to go through all the Pod and PodTemplate definitions and provide NodeAffinity boundaries to ensure safe scheduling requirements for the intended supported platforms, `amd64` and `arm64`.
  - Once filed, we should also clone a copy of this for LVMS :smile_cat: 
- As part of productization, we will externalize the manifest creation and bundling of kubesan into the kubesan repository. This will allow us to collect all of the manifests into a bundle container, which in turn would be used to create a catalog entry for our operator with all of its dependencies.
  - This PR injects those manifests into the operator directly to simplify the process of getting this up and running. It should be relatively easy to inject those from an external repository once we have official development builds. Given where we are in productization, I think we should move forward with this approach just to unblock development.
- Finally, we should probably do a security audit of our privileged pods in both KSAN and LVMS with the context of [trampoline pods](https://www.youtube.com/watch?v=qRo1Qw_Hr2A) in mind.


In short - if we create tickets to move forward these objectives, then we should be clear to `/unhold`

/lgtm
